File: Python/Max Points on a Line.py
# Definition for a point
# class Point:
#     def __init__(self, a=0, b=0):
#         self.x = a
#         self.y = b

class Solution:
    # @param points, a list of Points
    # @return an integer
    def maxPoints(self, points):
        ret = 0
        if len(points) < 2:
            return len(points)
        
        for i in range(len(points)):
            dupPoint = 1
            map = {}
            for j in range(i+1, len(points)):
                if points[i].x == points[j].x and points[i].y == points[j].y:
                    dupPoint += 1
                elif points[i].x == points[j].x:
                    self.helper(float('-inf'), map)
                # elif points[i].y == points[j].y:
                #     self.helper("horizon", map)
                else:
                    bias =(float(points[i].y - points[j].y)/float(points[i].x - points[j].x))
                    self.helper(bias, map)
            if map.values():
                maxI = max(map.values()) + dupPoint
            else:
                maxI = dupPoint
            ret = max(ret, maxI)
        return ret
    def helper(self, key, map):
        if key in map:
            map[key] += 1
        else:
            map[key] = 1

File: Java/src/Max Points on a Line.java
/**
 * Definition for a point.
 * class Point {
 *     int x;
 *     int y;
 *     Point() { x = 0; y = 0; }
 *     Point(int a, int b) { x = a; y = b; }
 * }
 */
public class Solution {
    public int maxPoints(Point[] points) {
        if (points.length < 2) return points.length;
        int ret = 0;
        for (int i = 0; i < points.length; i++){
            int dupPoint = 1;
            Map<Double, Integer> map = new HashMap<>();
            for (int j = i +1; j < points.length; j++){
            	if (points[i].x == points[j].x && points[i].y == points[j].y) {
					dupPoint++;
				}
            	else if (points[i].y == points[j].y) {
					helper(Double.NaN, map);
				}
            	else {
					double bias = Math.round( ((double)( points[i].x - points[j].x )/(double)( points[i].y - points[j].y)) * 1000 ) / 1000.0;
					helper(bias, map);
				}
            }
            
            int localMax = 0;
            for (int v : map.values()){
            	localMax = Math.max(localMax, v);
            }
            ret = Math.max(ret, localMax + dupPoint);
        }
        return ret;
    }
    
    private void helper(double key, Map<Double, Integer> map){
    	if (map.containsKey(key)) {
			map.put(key, map.get(key)+1);
		}
    	else {
			map.put(key, 1);
		}
    }
}